### PR TITLE
dev/core#3173 recaptcha support for formbuilder

### PR DESCRIPTION
**Pull request by kurund**
Overview
----------------------------------------
https://lab.civicrm.org/dev/core/-/issues/3173

### CODE REVIEW

**Comment by seamuslee001** [comment]
@kurund this is not about the code you have written which looks about right to me but I'm wondering if this should 'belong' in the recaptcha core extension or in afform. I can see arguments for both but I think the recaptcha one might be better in a way but I am open to discussion about where this should sit @colemanw @mattwire do either of you have any opinions on this?

**Comment by mattwire** [comment]
@colemanw Do we have any way to provide a "plugin" interface for afform? If possible it seems sensible to keep the recaptcha support separate to the afform extension to allow for privacy concerns / ability to put other solutions easily.

**Comment by colemanw** [comment]
> @colemanw Do we have any way to provide a "plugin" interface for afform? 

Yes we do. Any extension can modify the list of available elements via this event:
https://github.com/civicrm/civicrm-core/blob/0464c24e430a04b793fd9e1fe57180513b9fa6d6/ext/afform/admin/Civi/AfformAdmin/AfformAdminMeta.php#L285-L295

**Comment by colemanw** [comment]
@kurund I think the FormBuilder AdminUI part of this is less important than getting the actual Recaptha to work on front-end afforms, so let's not do any more AdminUI tweaks until the front-end part is working.

**Comment by kurund** [comment]
@colemanw 

Ready for your review.

**Comment by mattwire** [comment]
I'd really like to see this in a separate extension to afform as that makes it much easier to use alternative solutions and does not force you to use a google product.

**Comment by kurund** [comment]
@mattwire agree it would be good to support other solutions.  May be next phase?

**Comment by mattwire** [comment]
> @mattwire agree it would be good to support other solutions. May be next phase?

It's more about not having it as part of core. The work I did on moving ReCaptcha to a core extension and then the work I did on the formprotection extension to support ReCaptcha v3 were funded by German and US organisations who had reasons for specifically excluding ReCaptcha from core.
The main reason the core ReCaptcha extension is still hidden is because we never came up with an agreed interface for invoking it (core forms still call CRM_Utils_ReCaptcha to put it on forms). If we add this now to afform extension then we are giving ourselves more work to separate it later.
What I'm requesting really is that we find a way to include this in the core ReCaptcha extension and not the afform extension. That way we solve the problem of implementing other solutions and the interface at the same time - it seems like you should be able to use the `civi.afform_admin.metadata` event that @colemanw mentioned above for this?

**Comment by kurund** [comment]
> What I'm requesting really is that we find a way to include this in the core ReCaptcha extension and not the afform extension. That way we solve the problem of implementing other solutions and the interface at the same time - it seems like you should be able to use the civi.afform_admin.metadata event that @colemanw mentioned above for this?

I am not sure if it's possible in the current implementation. May be @colemanw know more on this?
- Recaptcha is added as a field element, supports v2 version as as core. Is it possible for an extension to add fields?
Screenshot from 2022-10-28 09-51-13
- Validation is done at client and also at the server level. Potentially, client side validation might be possible via ext. not sure about server side.

**Comment by mattwire** [comment]
In my extension here you can see that the extension chooses whether to add v2 or v3: https://lab.civicrm.org/extensions/formprotection/-/blob/main/CRM/Utils/ReCAPTCHA.php#L75 and then provides it's own form validation here: https://lab.civicrm.org/extensions/formprotection/-/blob/main/Civi/Formprotection/Recaptcha.php#L54

Of course I realise this is still quickform :-) but we have the opportunity to build things properly from the start with afform.

**Comment by colemanw** [comment]
I think it should be possible to move this code outside of the afform extension. I'll take a look at doing that.

File: ext/afform/core/ang/af/afRecaptcha.component.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+(function(angular, $, _) {
+  angular.module('af').component('afRecaptcha', {
+    templateUrl: '~/af/afRecaptcha.html',
+    controller: function($element) {
+      $('div.g-recaptcha').attr('data-sitekey', $($element).attr('recaptchakey'));
```
Comment (colemanw) [inline ext/afform/core/ang/af/afRecaptcha.component.js:5]: @kurund this is "the jQuery way" of doing things, where you manipulate the dom directly. To do this the angular way, you can keep the MetadataInjector code the same (where it sets the attribute `recaptchakey`) but you can use Angular to read it. Like this:
```suggestion
    bindings: {
      recaptchakey: '@',
    },
    controller: function($element) {
```

File: ext/afform/core/ang/af/afRecaptcha.html [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<script src="https://www.google.com/recaptcha/api.js" async defer></script>
+<div class="crm-af-field">
+  <div class="g-recaptcha"></div>
```
Comment (colemanw) [inline ext/afform/core/ang/af/afRecaptcha.html:3]: And then here you can insert it like this:
```suggestion
  <div class="g-recaptcha" data-sitekey="{{:: $ctrl.recaptchakey }}"></div>
```

File: ext/afform/core/ang/af/afRecaptcha.html [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<script src="https://www.google.com/recaptcha/api.js" async defer></script>
+<div class="crm-af-field">
```
Comment (colemanw) [inline ext/afform/core/ang/af/afRecaptcha.html:2]: This div doesn't seem to belong here. This isn't a field so it just seems confusing. Is it needed for some reason?

**Comment by colemanw** [comment]
Refactored into extension code as https://github.com/civicrm/civicrm-core/pull/24860

**Pull request closed without merging**